packages/go_router/lib/src/route.dart

```diff
@@ -458,4 +540,693 @@ class ShellRoute extends RouteBase {
+  }
+
+  @override
+  Object? toPrimitives() {
+    if (value != null && value.isNotEmpty) {
+      return _matchListCodec.encodeMatchList(value);
+    }
+    return null;
+  }
+}
+
+typedef _NavigatorForBranch = Widget? Function(StackedShellBranch);
+
+/// Widget that serves as the proxy for a branch Navigator Widget, which
+/// possibly hasn't been created yet.
+///
+/// This Widget hides the logic handling whether a Navigator Widget has been
+/// created yet for a branch or not, and at the same time ensures that the same
+/// Widget class is consistently passed to the containerBuilder. The latter is
+/// important for container implementations that cache child widgets,
+/// such as [TabBarView].
+class _BranchNavigatorProxy extends StatefulWidget {
+  const _BranchNavigatorProxy({
+    super.key,
+    required this.branch,
+    required this.navigatorForBranch,
+  });
+
+  final StackedShellBranch branch;
+  final _NavigatorForBranch navigatorForBranch;
+
+  @override
+  State<StatefulWidget> createState() => _BranchNavigatorProxyState();
+}
+
+/// State for _BranchNavigatorProxy, using AutomaticKeepAliveClientMixin to
+/// properly handle some scenarios where Slivers are used to manage the branches
+/// (such as [TabBarView]).
+class _BranchNavigatorProxyState extends State<_BranchNavigatorProxy>
+    with AutomaticKeepAliveClientMixin {
+  @override
+  Widget build(BuildContext context) {
+    super.build(context);
+    return widget.navigatorForBranch(widget.branch) ?? const SizedBox.shrink();
+  }
+
+  @override
+  bool get wantKeepAlive => true;
+}
+
+/// Default implementation of a container widget for the [Navigator]s of the
+/// route branches. This implementation uses an [IndexedStack] as a container.
+class _IndexedStackedRouteBranchContainer extends StatelessWidget {
+  const _IndexedStackedRouteBranchContainer(
+      {required this.currentIndex, required this.children});
+
+  final int currentIndex;
+
+  final List<Widget> children;
+
+  @override
+  Widget build(BuildContext context) {
+    final List<Widget> stackItems = children
+        .mapIndexed((int index, Widget child) =>
+            _buildRouteBranchContainer(context, currentIndex == index, child))
+        .toList();
+
+    return IndexedStack(index: currentIndex, children: stackItems);
+  }
+
+  Widget _buildRouteBranchContainer(
+      BuildContext context, bool isActive, Widget child) {
+    return Offstage(
+      offstage: !isActive,
+      child: TickerMode(
+        enabled: isActive,
+        child: child,
+      ),
+    );
+  }
+}
+
+/// The snapshot of the current state of a [StackedShellRoute].
+///
+/// Note that this an immutable class, that represents the snapshot of the state
+/// of a StackedShellRoute at a given point in time. Therefore, instances of
+/// this object should not be cached, but instead passed down from the builder
+/// functions of StackedShellRoute.
+@immutable
+class StackedShellRouteState {
```

Comment:
I'm not sure how much can be realistically re-used though. For instance, can we be sure `currentIndex` will be meaningful for other implementations of stateful shell routes? And if that field were to be removed, how would you access the current index before the shell is built the first time? You can't use an `.of(BuildContext)` method for this. 
 
Also, regarding moving goBranch, it feels odd to have different places for getting (`currentIndex`) and setting (`goBranch`) the index. I think perhaps a better solution would be to have different state classes for different stateful shell route implementations, possibly using a common base class. Or perhaps better yet, simply remove this class and use StackedNavigationShell directly (this is the direction I'm exploring right now - will check in a new version soon).

Reply:
>  can we be sure currentIndex will be meaningful for other implementations of stateful shell routes?

I imagine this can be reuse if they keep the branch like API and just using different container builder, such as tabview, navigation rail.

If they would implement a complete different way of doing stateful widget, they should extend ShellRouteBase instead